src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -234,6 +240,23 @@ private void WatchForFileSystemEventsThreadStart(object arg)
                 {
                     bool started = Interop.EventStream.FSEventStreamStart(_eventStream);
 
+                    _rootHandle = Interop.Sys.Open(_fullDirectory, Interop.Sys.OpenFlags.O_RDONLY, 0);
+                    if (_rootHandle.IsInvalid)
+                    {
+                        throw new FileNotFoundException(SR.Format(SR.FSW_IOError, _fullDirectory));
+                    }
+
+                    // Match Windows and lock the root directory
+                    // NOTE: this WILL NOT prevent this directory from being deleted from underneath us
+                    //       due to *nix having advisory-only locks (by default); this means that applications 
+                    //       can choose to ignore locks and operate on the FD anyway.
+                    if (Interop.Sys.FLock(_rootHandle, 
+                                          Interop.Sys.LockOperations.LOCK_EX | Interop.Sys.LockOperations.LOCK_NB) < 0)
+                    {
+                        // Match Windows and throw FileNotFoundException
+                        throw new FileNotFoundException(SR.Format(SR.FSW_IOError, _fullDirectory));
+                    }
```

Comment:
Since this is the same for both Linux and OS X, could you put it into a helper they both share, parameterized on the handle and the path?